### PR TITLE
[ota] OTA Image header parsing

### PR DESCRIPTION
**Pull request by Damian-Nordic**
#### Problem
We need a function for parsing OTA image header to be used by both OTA Provider (for example, to verify image integrity) and OTA Requestor (to validate the downloaded and determine the image size to be able to calculate the progress).

#### Change overview
Add a function for parsing OTA image header.

#### Testing
Added unit tests.

### CODE REVIEW

**Comment by selissia** [comment]
`DecodeOTAImageHeader()` should only be invoked on the full header, otherwise it will return an error. In practice the header will probably always fit in the first transfer block but the file-receiving Requestor code cannot make that assumption (I can't find anything in the spec that says otherwise). So the Requestor would need to make a determination whether the received image block contains the full header -- for that it would be nice to have a separate "peek" API whose output would be (1) TRUE if the input buffer contains an image header (i.e. magic value is there) and (2) the headerSize value (maybe also (3) the totalSize value).

**Review by selissia — APPROVED** [review]

**Comment by Damian-Nordic** [comment]
@selissia I made the function to return `CHIP_ERROR_BUFFER_TOO_SMALL` if the buffer doesn't contain the full header (although `CHIP_ERROR_INVALID_FILE_IDENTIFIER` can be returned if just 4 bytes are available and are invalid to quickly discard non-compliant files), so the requestor will be able to accumulate subsequent blocks until another error or no error is returned. That's what I'm planning to do in `BDXDownloader`.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by LuDuda — APPROVED** [review]

File: src/lib/core/OTAImageHeader.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+struct OTAImageHeader
+{
+    uint64_t mTotalSize;
+    uint32_t mHeaderSize;
+    uint16_t mVendorId;
+    uint16_t mProductId;
+    uint32_t mSoftwareVersion;
+    CharSpan mSoftwareVersionString;
```
Comment (tcarmelveilleux) [inline src/lib/core/OTAImageHeader.h:52]: Please add comment that the struct does not own the bytes for all spans, and therefore these values must be copied if they are to be used after the header is read.

File: src/lib/core/OTAImageHeader.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    ReturnErrorOnFailure(tlvReader.EnterContainer(outerType));
+
+    ReturnErrorOnFailure(tlvReader.Next(TLV::ContextTag(to_underlying(Tag::kVendorId))));
+    ReturnErrorOnFailure(tlvReader.Get(header.mVendorId));
+    ReturnErrorOnFailure(tlvReader.Next(TLV::ContextTag(to_underlying(Tag::kProductId))));
+    ReturnErrorOnFailure(tlvReader.Get(header.mProductId));
+    ReturnErrorOnFailure(tlvReader.Next(TLV::ContextTag(to_underlying(Tag::kSoftwareVersion))));
+    ReturnErrorOnFailure(tlvReader.Get(header.mSoftwareVersion));
+    ReturnErrorOnFailure(tlvReader.Next(TLV::ContextTag(to_underlying(Tag::kSoftwareVersionString))));
+    ReturnErrorOnFailure(tlvReader.Get(header.mSoftwareVersionString));
```
Comment (tcarmelveilleux) [inline src/lib/core/OTAImageHeader.cpp:71]: Suggest erroring out if string length is > 64.

**Pull request merged**